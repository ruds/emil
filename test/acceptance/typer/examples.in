(* Examples from various other places. *)

(* The next few are from
 * https://www.smlnj.org/doc/Conversion/types.html
 *)

(* dummy definitions *)
val ar = fn (_ : 'a ref, a : 'a) => ref a;
val dr = fn (ref a) => a;
val hd = fn ((::) (a, _)) => a;

(* Expansive: can't generalize *)
val x = ref nil;
val x : int list ref = ref nil;

(* Unifies *)
let val x = ref nil
 in ar (x, [3]);
    hd (dr x)
end;

(* Generalizes *)
val f = fn y => let val x = ref nil in x end;

(* Instantiates *)
let val y = let val x = ref nil in x end
 in ar (y, [true]);
    hd (dr y)
end;

val f = fn x => fn y => y;
val g = f 3;
val g = (fn y => (f 3) y);
val g = fn y => f 3 y;

val f = fn ((::) (x, l)) => (fn y => y);
val g = f nil;
val g = fn y => f nil y;

val counter = ref(ref 0);
val mkCountF = fn f =>
    let val x = ref 0
        val f2 = fn z => (ar (x, dr x); f z)
     in ar (counter, x);
        f2
     end;
val pair = fn x => (x, x);
val cpair = mkCountF pair;
val cpairCalls = dr counter;
val cpair' = (fn y => mkCountF pair y);

let val r = ref(fn x => x)
 in (dr r) 5
end;

let val r = ref(fn x => x)
 in (dr r) 5; (dr r) true
end;

val f = fn () =>
    let val r = ref(fn x => x)
     in (dr r)
    end;

val x = ref(fn x => x);

val x = (let val id : 'a -> 'a = fn z => z in id id end,
         fn y => y);
val x = (let val id : 'a -> 'a = fn z => z in id id end,
         fn (y: 'a) => y);
(* TODO: add explicit type variable binding examples *)
