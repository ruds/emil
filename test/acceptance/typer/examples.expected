@0008:
val (:=) : ('a ref * 'a) -> 'a ref
@0009:
val (prefix !) : 'a ref -> 'a
@0010: WARNING
@0010:
val hd : 'a list -> 'a
@0013: WARNING
@0013:
val x : X1 list ref
@0014:
val x : int list ref
@0017:
val it : bigint
@0023:
val f : 'a -> 'b list ref
@0026:
val it : bool
@0031:
val f : 'a -> 'b -> 'b
@0032: WARNING
@0032:
val g : X1 -> X1
@0033:
val g : 'a -> 'a
@0034:
val g : 'a -> 'a
@0036: WARNING
@0036:
val f : 'a list -> 'b -> 'b
@0037: WARNING
@0037:
val g : X1 -> X1
@0038:
val g : 'a -> 'a
@0040:
val counter : bigint ref ref
@0041:
val mkCountF : ('a -> 'b) -> 'a -> 'b
@0047:
val pair : 'a -> ('a * 'a)
@0048: WARNING
@0048:
val cpair : X1 -> (X1 * X1)
@0049:
val cpairCalls : bigint ref
@0050:
val cpair' : 'a -> ('a * 'a)
@0052:
val it : bigint
@0057: ERROR
@0060:
val f : () -> 'a -> 'a
@0065: WARNING
@0065:
val x : (X1 -> X1) ref
@0067: WARNING
@0067:
val x : ((X1 -> X1) * (X2 -> X2))
@0069: ERROR
@0071: WARNING
@0071:
val x : X1 -> X1
@0072: ERROR
@0073: ERROR
@0074: ERROR
@0077:
datatype t1 = C1
val C1 : t1
@0078:
val it : t1
@0080: ERROR
@0081: ERROR
@0084: ERROR
@0091: ERROR
@0098: ERROR
@0103:
val f : 'a -> 'a
@0113:
val it : t1 list ref
@0118: ERROR
@0120:
